core/src/main/java/cucumber/runtime/xstream/ComplexTypeWriter.java

```diff
@@ -30,24 +32,24 @@ public ComplexTypeWriter(List<String> columnNames) {
             String[] explicitFieldValues = new String[columnNames.size()];
             int n = 0;
             for (String columnName : columnNames) {
-                int index = fieldNames.indexOf(converter.map(columnName));
-                if (index == -1) {
-                    explicitFieldValues[n] = "";
+                final String convertedColumnName = converter.map(columnName);
+                if (fields.containsKey(convertedColumnName)) {
+                    explicitFieldValues[n] = fields.get(convertedColumnName);
                 } else {
-                    explicitFieldValues[n] = fieldValues.get(index);
+                    explicitFieldValues[n] = "";
                 }
                 n++;
             }
             return asList(explicitFieldValues);
         } else {
-            return fieldValues;
+            return new ArrayList<String>(fields.values());
         }
     }
 
     @Override
     public void startNode(String name) {
         if (nodeDepth == 1) {
-            this.fieldNames.add(name);
+            currentKey = name;
```

Comment:
Ok, I can take a look at fixing this! Since the current tests all pass I think some more are needed. Would it be possible for you to add one or more failing tests that show the behaviour? I can try adding them myself but I don't have enough knowledge of all the various use cases to support.